broadcast/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//! Replication of messages across a network.
+
+use bytes::Bytes;
+use commonware_cryptography::Array;
+use futures::channel::oneshot;
+use std::future::Future;
+
+pub mod linked;
+
+/// Proof is a blob that attests to some data.
+pub type Proof = Bytes;
+
+/// Broadcaster is the interface responsible for replication of messages across a network.
+pub trait Broadcaster: Clone + Send + 'static {
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Attempt to broadcast a digest to the network.
+    ///
+    /// Returns a future that resolves to a boolean indicating success.
+    /// The broadcast may fail for a variety of reasons such-as networking errors, the node not
+    /// being a valid sequencer, or the Broadcaster not being ready to broadcast a new payload.
+    fn broadcast(
+        &mut self,
+        payload: Self::Digest,
+    ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
+}
+
+/// Application is the interface responsible for processing messages received from the network.
+pub trait Application: Clone + Send + 'static {
+    /// Context is metadata provided by the broadcast engine to associated with a given payload.
+    /// This could include things like the public key of the sequencer.
+    type Context;
+
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Verify a proposed payload received from the network.
+    ///
+    /// Returns a future that resolves to a boolean indicating success.
+    /// Part of verification requires ensuring that the data is made available.
+    /// For example, by storing it in a database.
+    fn verify(
+        &mut self,
+        context: Self::Context,
+        payload: Self::Digest,
+    ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
+}
+
+/// Collector is the interface responsible for handling notifications of broadcasted payloads.
+pub trait Collector: Clone + Send + 'static {
+    /// Context is metadata provided by the broadcast engine to associated with a given payload.
+    /// This could include things like the public key of the sequencer.
+    type Context;
+
+    /// Digest is an arbitrary hash digest.
+    type Digest: Array;
+
+    /// Emit that a payload has been successfully broadcasted.
+    /// This is used to acknowledge that the payload has been "received" by the network,
+    /// for example that it has been successfully gossiped to a threshold of validators.
+    fn acknowledged(
+        &mut self,
+        context: Self::Context,
```

Comment:
Could you push the change that removes context here?

Reply:
Oops yeah my bad